### DISCUSSION TITLE
How to call a non-public variable within a contract?

### DISCUSSION THREAD

**Discussion by andreachello** [ethereumjs/ethereumjs-monorepo#3782] [Q&A] [answered]
This example

```js
async function getGreeting(vm: VM, contractAddress: Address, caller: Address) {
  const sigHash = new Interface(['function greet()']).getSighash('greet')

  const greetResult = await vm.evm.runCall({
    to: contractAddress,
    caller: caller,
    origin: caller, // The tx.origin is also the caller here
    data: hexToBytes(sigHash),
    block,
  })

  if (greetResult.execResult.exceptionError) {
    throw greetResult.execResult.exceptionError
  }

  const results = AbiCoder.decode(['string'], greetResult.execResult.returnValue)

  return results[0]
}
```

doesn't work if instead of using a function i try to put in a variable name. How can I call a variable?

**Comment by jochem-brouwer** [comment]
I don't understand what you mean by "putting a variable name", could you post an example or description what you are trying to do? :smile:

**Reply by andreachello** [reply]
Yeah, sorry should have clarified, I mean if you have a variable in a contract like: 

```solidity
contract Example {
    uint public a = 12;
}
```

in order to get the value in the variable we can do 

```js
const sigHash = new Interface(['function a()']).getSighash('a')
```

only if it is a `public` variable. Is there a way to get the value of the variable even without declaring it as a public variable, like getStorageAt?

**Reply by jochem-brouwer** [reply, marked as answer]
No, you cannot do that via `evm.runCall`. It is not possible to read "private" variables within the EVM. But, it is indeed possible outside the EVM via `getStorageAt` RPC calls.
